cmd/dynamic_startup.go

```diff
@@ -0,0 +1,106 @@
+package cmd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/onflow/flow-go/cmd/util/cmd/common"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol/inmem"
+	"github.com/rs/zerolog"
+)
+
+const sleepInterval = time.Minute
+
+func DynamicStartup(logger zerolog.Logger, initANAddress, initANPubkey string, startupEpoch uint64, startupEpochPhase flow.EpochPhase, rootSnapshotOutDir string) error {
+
+	// get flow client config that connects to our trusted access node
+	config, err := common.NewFlowClientConfig(initANAddress, initANPubkey, false)
+	if err != nil {
+		return fmt.Errorf("failed to create flow client config for node dynamic startup pre-init: %w", err)
+	}
+
+	flowClient, err := common.FlowClient(config)
+	if err != nil {
+		return fmt.Errorf("failed to create flow client for node dynamic startup pre-init: %w", err)
+	}
+
+	ctx := context.Background()
+	start := time.Now()
+	var snapshot *inmem.Snapshot
+
+	// wait until we reach the desired Epoch and EpochPhase
+	for {
+		b, err := flowClient.GetLatestProtocolStateSnapshot(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get latest finalized protocol state snapshot during pre-initialization: %w", err)
```

Comment:
are any of these errors transient/might recover without a restart?

Reply:
https://github.com/onflow/flow-go/pull/1893/commits/8ea656ac507e9969c635a600e908891394d7ac2c